File: backend/app/api/v1/mappings.py
"""Mapping endpoints."""

import uuid
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.control import ControlMapping
from app.models.policy import PolicyMapping
from app.models.user import User
from app.schemas.mapping import (
    MappingGenerateRequest,
    MappingGenerateResponse,
    MappingApproveRequest,
    GapListResponse,
    BulkMappingRequest,
    BulkMappingResponse,
)
from app.schemas.control import ControlMappingResponse
from app.schemas.policy import PolicyMappingResponse
from app.dependencies.auth import get_current_user, require_user
from app.services.mapping.ai_mapper import AIMappingService
from app.services.mapping.gap_detector import GapDetectionService

router = APIRouter()


@router.post(
    "/assessments/{assessment_id}/generate",
    response_model=MappingGenerateResponse,
)
async def generate_mappings(
    assessment_id: uuid.UUID,
    request: MappingGenerateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user),
):
    """Generate AI-powered mapping suggestions for an assessment."""
    mapper = AIMappingService(db)

    result = mapper.generate_mappings_for_assessment(
        assessment_id=assessment_id,
        user_id=current_user.id,
        include_policies=request.include_policies,
        include_controls=request.include_controls,
        confidence_threshold=request.confidence_threshold,
    )

    return MappingGenerateResponse(**result)


@router.get("/assessments/{assessment_id}/policies")
async def get_policy_mappings(
    assessment_id: uuid.UUID,
    approved_only: bool = False,
    db: Session = Depends(get_db),
    current_user: User | None = Depends(get_current_user),
):
    """Get all policy mappings for an assessment."""
    from app.models.policy import Policy

    query = (
        db.query(PolicyMapping)
        .join(Policy)
        .filter(Policy.assessment_id == assessment_id)
    )

    if approved_only:
        query = query.filter(PolicyMapping.is_approved == True)

    mappings = query.all()

    return [PolicyMappingResponse.model_validate(m) for m in mappings]


@router.get("/assessments/{assessment_id}/controls")
async def get_control_mappings(
    assessment_id: uuid.UUID,
    approved_only: bool = False,
    db: Session = Depends(get_db),
    current_user: User | None = Depends(get_current_user),
):
    """Get all control mappings for an assessment."""
    from app.models.control import Control

    query = (
        db.query(ControlMapping)
        .join(Control)
        .filter(Control.assessment_id == assessment_id)
    )

    if approved_only:
        query = query.filter(ControlMapping.is_approved == True)

    mappings = query.all()

    return [ControlMappingResponse.model_validate(m) for m in mappings]


@router.post("/{mapping_id}/approve")
async def approve_mapping(
    mapping_id: uuid.UUID,
    mapping_type: Literal["policy", "control"],
    request: MappingApproveRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user),
):
    """Approve or reject a mapping suggestion."""
    mapper = AIMappingService(db)

    result = mapper.approve_mapping(
        mapping_id=mapping_id,
        mapping_type=mapping_type,
        approved=request.is_approved,
        user_id=current_user.id,
    )

    if "error" in result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=result["error"],
        )

    return result


@router.post("/{mapping_id}/reject")
async def reject_mapping(
    mapping_id: uuid.UUID,
    mapping_type: Literal["policy", "control"],
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user),
):
    """Reject a mapping suggestion."""
    mapper = AIMappingService(db)

    result = mapper.approve_mapping(
        mapping_id=mapping_id,
        mapping_type=mapping_type,
        approved=False,
        user_id=current_user.id,
    )

    if "error" in result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=result["error"],
        )

    return result


@router.get("/assessments/{assessment_id}/gaps", response_model=GapListResponse)
async def get_gaps(
    assessment_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User | None = Depends(get_current_user),
):
    """Get coverage gaps for an assessment."""
    gap_detector = GapDetectionService(db)
    result = gap_detector.detect_gaps(assessment_id)

    return GapListResponse(**result)


@router.post("/bulk-approve", response_model=BulkMappingResponse)
async def bulk_approve_mappings(
    request: BulkMappingRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user),
):
    """
    Bulk approve multiple mappings.

    Approves all valid mappings and reports any failures.
    """
    from datetime import datetime
    from app.services.audit.audit_service import AuditService

    results = []
    successful = 0
    failed = 0

    audit_service = AuditService(db)

    for mapping_id in request.mapping_ids:
        try:
            if request.mapping_type == "control":
                mapping = db.query(ControlMapping).filter(
                    ControlMapping.id == mapping_id
                ).first()
            else:
                mapping = db.query(PolicyMapping).filter(
                    PolicyMapping.id == mapping_id
                ).first()

            if not mapping:
                results.append({
                    "mapping_id": mapping_id,
                    "success": False,
                    "error": "Mapping not found",
                })
                failed += 1
                continue

            mapping.is_approved = True
            mapping.approved_by_id = current_user.id
            mapping.approved_at = datetime.utcnow()

            audit_service.log_update(
                entity_type=f"{request.mapping_type}_mapping",
                entity_id=mapping_id,
                old_values={"is_approved": False},
                new_values={"is_approved": True},
                user_id=current_user.id,
            )

            results.append({
                "mapping_id": mapping_id,
                "success": True,
                "error": None,
            })
            successful += 1

        except Exception as e:
            results.append({
                "mapping_id": mapping_id,
                "success": False,
                "error": str(e),
            })
            failed += 1

    db.commit()

    return BulkMappingResponse(
        total=len(request.mapping_ids),
        successful=successful,
        failed=failed,
        results=results,
    )


@router.post("/bulk-reject", response_model=BulkMappingResponse)
async def bulk_reject_mappings(
    request: BulkMappingRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user),
):
    """
    Bulk reject (delete) multiple mappings.

    Deletes all valid mappings and reports any failures.
    """
    from app.services.audit.audit_service import AuditService

    results = []
    successful = 0
    failed = 0

    audit_service = AuditService(db)

    for mapping_id in request.mapping_ids:
        try:
            if request.mapping_type == "control":
                mapping = db.query(ControlMapping).filter(
                    ControlMapping.id == mapping_id
                ).first()
            else:
                mapping = db.query(PolicyMapping).filter(
                    PolicyMapping.id == mapping_id
                ).first()

            if not mapping:
                results.append({
                    "mapping_id": mapping_id,
                    "success": False,
                    "error": "Mapping not found",
                })
                failed += 1
                continue

            db.delete(mapping)

            audit_service.log_delete(
                entity_type=f"{request.mapping_type}_mapping",
                entity_id=mapping_id,
                old_values={"is_approved": mapping.is_approved},
                user_id=current_user.id,
            )

            results.append({
                "mapping_id": mapping_id,
                "success": True,
                "error": None,
            })
            successful += 1

        except Exception as e:
            results.append({
                "mapping_id": mapping_id,
                "success": False,
                "error": str(e),
            })
            failed += 1

    db.commit()

    return BulkMappingResponse(
        total=len(request.mapping_ids),
        successful=successful,
        failed=failed,
        results=results,
    )

File: frontend/lib/api/mappings.ts
import { apiRequest } from './client';
import {
  MappingGenerateRequest,
  MappingGenerateResponse,
  MappingApproveResponse,
  GapListResponse,
  ControlMapping,
  PolicyMapping,
} from '../types';

export async function generateMappings(
  assessmentId: string,
  options: MappingGenerateRequest = {},
  userId?: string
): Promise<MappingGenerateResponse> {
  return apiRequest<MappingGenerateResponse>(
    `/mappings/assessments/${assessmentId}/generate`,
    {
      method: 'POST',
      body: options,
      userId,
    }
  );
}

export async function approveMapping(
  mappingId: string,
  userId?: string
): Promise<MappingApproveResponse> {
  return apiRequest<MappingApproveResponse>(
    `/mappings/${mappingId}/approve`,
    {
      method: 'POST',
      userId,
    }
  );
}

export async function rejectMapping(
  mappingId: string,
  userId?: string
): Promise<MappingApproveResponse> {
  return apiRequest<MappingApproveResponse>(
    `/mappings/${mappingId}/reject`,
    {
      method: 'POST',
      userId,
    }
  );
}

export async function listControlMappings(
  assessmentId: string,
  userId?: string
): Promise<ControlMapping[]> {
  return apiRequest<ControlMapping[]>(
    `/mappings/assessments/${assessmentId}/controls`,
    { userId }
  );
}

export async function listPolicyMappings(
  assessmentId: string,
  userId?: string
): Promise<PolicyMapping[]> {
  return apiRequest<PolicyMapping[]>(
    `/mappings/assessments/${assessmentId}/policies`,
    { userId }
  );
}

export async function getGaps(
  assessmentId: string,
  userId?: string
): Promise<GapListResponse> {
  return apiRequest<GapListResponse>(
    `/mappings/assessments/${assessmentId}/gaps`,
    { userId }
  );
}

export interface BulkMappingResult {
  mapping_id: string;
  success: boolean;
  error: string | null;
}

export interface BulkMappingResponse {
  total: number;
  successful: number;
  failed: number;
  results: BulkMappingResult[];
}

export async function bulkApproveMappings(
  mappingIds: string[],
  mappingType: 'policy' | 'control',
  userId?: string
): Promise<BulkMappingResponse> {
  return apiRequest<BulkMappingResponse>('/mappings/bulk-approve', {
    method: 'POST',
    body: { mapping_ids: mappingIds, mapping_type: mappingType },
    userId,
  });
}

export async function bulkRejectMappings(
  mappingIds: string[],
  mappingType: 'policy' | 'control',
  userId?: string
): Promise<BulkMappingResponse> {
  return apiRequest<BulkMappingResponse>('/mappings/bulk-reject', {
    method: 'POST',
    body: { mapping_ids: mappingIds, mapping_type: mappingType },
    userId,
  });
}
